### PR TITLE
[SPARK-46466][SQL] Vectorized parquet reader should never do rebase for timestamp ntz

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This fixes a correctness bug. The TIMESTAMP_NTZ is a new data type in Spark and has no legacy files that need to do calendar rebase. However, the vectorized parquet reader treat it the same as LTZ and may do rebase if the parquet file was written with the legacy rebase mode. This PR fixes it to never do rebase for NTZ.

### Why are the changes needed?

bug fix

### Does this PR introduce _any_ user-facing change?

Yes, now we can correctly write and read back NTZ value even if the date is before 1582.

### How was this patch tested?

new test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gengliangwang @MaxGekk @dongjoon-hyun

**Comment by MaxGekk** [comment]
@cloud-fan Thanks for the ping. I will review this PR tomorrow.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 1]
```diff
@@ -255,6 +255,18 @@ abstract class ParquetQuerySuite extends QueryTest with ParquetTest with SharedS
     }
   }
 
+  test("write and read TimestampNTZ with legacy rebase mode") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala:258]: If you don't mind, please add a test prefix, `SPARK-46466: ` here for easy tracing this correctness bug.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 2]
```diff
@@ -195,12 +203,11 @@ boolean isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit unit) {
       annotation.getUnit() == unit;
   }
 
-  void validateTimestampType(DataType sparkType) {
+  void validateTimestampNTZType(DataType sparkType) {
     assert(logicalTypeAnnotation instanceof TimestampLogicalTypeAnnotation);
-    // Throw an exception if the Parquet type is TimestampLTZ and the Catalyst type is TimestampNTZ.
+    // Throw an exception if the Parquet type is TimestampLTZ as the Catalyst type is TimestampNTZ.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:208]: This looks like a wrong change. Could you double-check?

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
I updated the JIRA as a correctness blocker issue for Apache Spark 3.5.1 and 3.4.3. Thank you for the fix.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala:258]:
```suggestion
  test("SPARK-46466: write and read TimestampNTZ with legacy rebase mode") {
```

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 3]
```diff
@@ -109,24 +109,32 @@ public ParquetVectorUpdater getUpdater(ColumnDescriptor descriptor, DataType spa
           // For unsigned int64, it stores as plain signed int64 in Parquet when dictionary
           // fallbacks. We read them as decimal values.
           return new UnsignedLongUpdater();
-        } else if (isTimestamp(sparkType) &&
```
Comment (beliefer) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:112]: Shall we remove `isTimestamp` is not used any more?

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:112]: done

**Review by MaxGekk — COMMENTED** [review]
> The TIMESTAMP_NTZ is a new data type in Spark and has no legacy files that need to do calendar rebase.

But the behaviour of rebasing of `TIMESTAMP_NTZ` has been released already by at least Spark 3.5 (and maybe 3.4). So, Spark users might write `TIMESTAMP_NTZ` with the rebase SQL config set to `LEGACY`. How will they read the data back after switching to Spar 4.0.0. Seems we need a legacy config to restore the previous behaviour, don't we?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 4]
```diff
@@ -195,12 +203,11 @@ boolean isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit unit) {
       annotation.getUnit() == unit;
   }
 
-  void validateTimestampType(DataType sparkType) {
+  void validateTimestampNTZType(DataType sparkType) {
```
Comment (LuciferYang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:206]: 1. The input parameter `sparkType` is no longer in use, can it be removed?
2. Although the old code is like this, in non-ea mode, `assert` is invalid.
3. Is it possible to reduce the access scope of the `validateTimestampNTZType` method to `private` in this pr? It does not need to be accessed by other classes.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java [hunk 5]
```diff
@@ -109,24 +109,32 @@ public ParquetVectorUpdater getUpdater(ColumnDescriptor descriptor, DataType spa
...
           if ("CORRECTED".equals(datetimeRebaseMode)) {
             return new LongAsMicrosUpdater();
           } else {
             final boolean failIfRebase = "EXCEPTION".equals(datetimeRebaseMode);
             return new LongAsMicrosRebaseUpdater(failIfRebase, datetimeRebaseTz);
           }
+        } else if (sparkType == DataTypes.TimestampNTZType &&
+          isTimestampTypeMatched(LogicalTypeAnnotation.TimeUnit.MICROS)) {
+          validateTimestampNTZType(sparkType);
+          // TIMESTAMP_NTZ is a new data type and has no legacy files that need to do rebase.
```
Comment (beliefer) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:131]: Just a question. `TIMESTAMP_NTZ` has been released at 3.5.0, why there has no legacy files?

**Review by beliefer — APPROVED** [review]
LGTM except one comment.

**Comment by cloud-fan** [comment]
> But the behaviour of rebasing of TIMESTAMP_NTZ has been released already by at least Spark 3.5 (and maybe 3.4). 

No, we never rebase ntz values during writing. It's a pure bug that we never rebase ntz during writing but may rebase during reading.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:131]: legacy here means parquet files written before the calendar switch.

**Review by LuciferYang — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @cloud-fan and @dongjoon-hyun @LuciferYang @beliefer for review.

**Comment by MaxGekk** [comment]
@cloud-fan BTW, this should be backported to 3.4 and 3.5, correct?

**Review by beliefer — APPROVED** [review]

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java (hunk 5 above)
Comment (beliefer) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java:131]: Got it now. You means we never rebase the time zone for `TIMESTAMP_NTZ`.

**Comment by gengliangwang** [comment]
Late LGTM! Thanks for the fix.

**Pull request closed without merging**